_source/_posts/2025-07-28-express-oauth-pkce.md

```diff
@@ -0,0 +1,700 @@
+      scope: "openid profile email offline_access",
+      state,
+      code_challenge,
+      code_challenge_method: "S256",
+      redirect_uri: `${process.env.APP_BASE_URL}/authorization-code/callback`,
+    });
+
+    res.redirect(authUrl);
+  } catch (error) {
+    res.status(500).send("Something failed during the authorization request");
+  }
+}
+
+function getCallbackUrlWithParams(req) {
+  const host = req.headers["x-forwarded-host"] || req.headers.host || "localhost";
+  const protocol = req.headers["x-forwarded-proto"] || req.protocol;
+  const currentUrl = new URL(`${protocol}://${host}${req.originalUrl}`);
+  return currentUrl;
+}
+
+export async function authCallback(req, res, next) {
+  try {
+    const openIdClientConfig = await getClientConfig();
+
+    const { pkce } = req.session;
+
+    if (!pkce || !pkce.code_verifier || !pkce.state) {
+      throw new Error("Login session expired or invalid. Please try logging in again.");
+    }
+
+    const tokenSet = await client.authorizationCodeGrant(openIdClientConfig, getCallbackUrlWithParams(req), {
+      pkceCodeVerifier: pkce.code_verifier,
+      expectedState: pkce.state,
+    });
+
+    const { name, email } = tokenSet.claims();
+    const teams = getModifiedTeam(userTeamMap[email]);
+
+    const userProfile = {
+      name,
+      email,
+      teams,
+      idToken: tokenSet.id_token,
+    };
+
+    delete req.session.pkce;
+
+    req.logIn(userProfile, (err) => {
+      if (err) {
+        return next(err);
+      }
+
+      return res.redirect("/dashboard");
+    });
+  } catch (error) {
+    console.error("Authentication error:", error.message);
+    return res.status(500).send(`Authentication failed: ${error.message}`);
+  }
+}
+
+export async function logout(req, res) {
+  try {
+    const openIdClientConfig = await getClientConfig();
+
+    const id_token_hint = req.user?.idToken;
+
+    const logoutUrl = client.buildEndSessionUrl(openIdClientConfig, {
+      id_token_hint,
+      post_logout_redirect_uri: process.env.POST_LOGOUT_URL,
+    });
+
+    req.logout((err) => {
+      if (err) return next(err);
+
+      req.session.destroy((err) => {
+        if (err) return next(err);
+        res.redirect(logoutUrl);
+      });
+    });
+  } catch (error) {
+    res.status(500).send('Something went wrong during logout.');
+  }
+}
+
+export function ensureAuthenticated(req, res, next) {
+  if (req.isAuthenticated()) {
+    return next();
+  }
+  res.redirect("/login");
+}
+```
+
+This file includes the following functions:
+
+1. **`getClientConfig`** - Discovers the authorization server's metadata and returns the associated metadata object.
+2. **`login`** - This function starts the Authorization Code + PKCE flow. It generates a **code_verifier**, derives a **code_challenge**, and creates a state value. It stores the **code_verifier** and **state** in the session, builds the authorize URL, and redirects the user to Okta to complete the login flow.
```

Comment:
```suggestion
2. **`login`** - This function starts the Authorization Code + PKCE flow. It generates the required values to enable PKCE: the **code_verifier** and **code_challenge**. These values, along with the `state` value protect the user sign in process from attack vectors. PKCE protects against auth code interception attacks and the state parameter protects against Cross-site Request Forgery (CSRF). The openid-client builds the user sign in URL with these values and redirects the user to Okta to complete the authentication challenge.
```

Explaining a bit more about why each value instead of rehashing the code. Might want to consider linking to PKCE resource and CSRF too?

Reply:
Not Grammarly checked